pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -609,3 +612,273 @@ func TestApplyWorkspaces(t *testing.T) {
+		results: []v1beta1.PipelineResult{{
+			Name:  "foo",
+			Value: "$(tasks.pt1.results.foo)",
+		}},
+		statuses: map[string]*v1beta1.PipelineRunTaskRunStatus{
+			"task1": {
+				PipelineTaskName: "pt1",
+				Status: &v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{
+							Type:   apis.ConditionSucceeded,
+							Status: corev1.ConditionTrue,
+						}},
+					},
+					TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+						TaskRunResults: []v1beta1.TaskRunResult{{
+							Name:  "definitely-not-foo",
+							Value: "bar",
+						}},
+					},
+				},
+			},
+		},
+		expected: nil,
+	}, {
+		description: "unsuccessful-taskrun-no-returned-result",
+		results: []v1beta1.PipelineResult{{
+			Name:  "foo",
+			Value: "$(tasks.pt1.results.foo)",
+		}},
+		statuses: map[string]*v1beta1.PipelineRunTaskRunStatus{
+			"task1": {
+				PipelineTaskName: "pt1",
+				Status: &v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{
+							Type:   apis.ConditionSucceeded,
+							Status: corev1.ConditionFalse,
+						}},
+					},
+					TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+						TaskRunResults: []v1beta1.TaskRunResult{{
+							Name:  "foo",
+							Value: "bar",
+						}},
+					},
+				},
+			},
+		},
+		expected: nil,
+	}, {
+		description: "mixed-success-tasks-some-returned-results",
+		results: []v1beta1.PipelineResult{{
+			Name:  "foo",
+			Value: "$(tasks.pt1.results.foo)",
+		}, {
+			Name:  "bar",
+			Value: "$(tasks.pt2.results.bar)",
+		}},
+		statuses: map[string]*v1beta1.PipelineRunTaskRunStatus{
+			"task1": {
+				PipelineTaskName: "pt1",
+				Status: &v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{
+							Type:   apis.ConditionSucceeded,
+							Status: corev1.ConditionFalse,
+						}},
+					},
+					TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+						TaskRunResults: []v1beta1.TaskRunResult{{
+							Name:  "foo",
+							Value: "do",
+						}},
+					},
+				},
+			},
+			"task2": {
+				PipelineTaskName: "pt2",
+				Status: &v1beta1.TaskRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{
+							Type:   apis.ConditionSucceeded,
+							Status: corev1.ConditionTrue,
+						}},
+					},
+					TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+						TaskRunResults: []v1beta1.TaskRunResult{{
+							Name:  "bar",
+							Value: "rae",
+						}},
+					},
+				},
+			},
+		},
+		expected: []v1beta1.PipelineRunResult{{
+			Name:  "bar",
+			Value: "rae",
+		}},
+	}, {
+		description: "multiple-results-multiple-successful-tasks",
+		results: []v1beta1.PipelineResult{{
+			Name:  "pipeline-result-1",
+			Value: "$(tasks.pt1.results.foo)",
+		}, {
+			Name:  "pipeline-result-2",
+			Value: "$(tasks.pt1.results.foo), $(tasks.pt2.results.baz), $(tasks.pt1.results.bar), $(tasks.pt2.results.baz), $(tasks.pt1.results.foo)",
```

Comment:
@sbwsg while your are updating the doc, might be worth mentioning this particular scenario.

I think without these changes (this PR):

```
"$(tasks.pt1.results.foo), $(tasks.pt2.results.baz), $(tasks.pt1.results.bar), $(tasks.pt2.results.baz), $(tasks.pt1.results.foo)"
```

if `$(tasks.pt2.results.baz)` was not initialized, this was resulting in:

```
"do, $(tasks.pt2.results.baz), mi, $(tasks.pt2.results.baz), do"
```

(I haven't confirmed yet)

But now with these changes, its not set at all and not part of the results any more.

Reply:
yup just confirmed, on master, its producing "do, $(tasks.pt2.results.baz), mi, $(tasks.pt2.results.baz), do" 😢

Reply:
Thanks @pritidesai 👍 right, that's the bad behaviour we're aiming to fix here.

Reply:
I've added a Note after the list of reasons that a `Pipeline Result` will not be emitted.  The note talks about a Pipeline Result that references multiple Taskrun Results and makes clear that any invalid reference from the Pipeline Result will mean the entire result is not emitted.